PR: Restore widget shortcuts to Preferences and allow to change them on the fly (Shortcuts)

ccordoba12 opened the pull request:
## Description of Changes

- Those shortcuts were not shown in Spyder 6, as reported by several users.
- That was a regression introduced by #22230, which unfortunately we overlooked.
- This also fixes an important usability issue: changing those shortcuts didn't have an immediate effect (like all other shortcuts), but required an app restart. That issue was also present in Spyder 5 and it was very counterintuitive. Now those shortcuts will be updated on the fly.
- Introduce a new `ShortcutData` dataclass to represent shortcut data internally and work more easily with it.
- Change context of Switcher plugin shortcuts to be `_`, i.e. global (before their context was `switcher` but that made little sense).

### Issue(s) Resolved

Fixes #23072 
Fixes #22741 
Fixes #22516

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

<!--- TYPE YOUR USER/NAME AFTER THE FOLLOWING: --->
I certify the above statement is true and correct: @ccordoba12 

<!--- Thanks for your help making Spyder better for everyone! --->

ccordoba12 commented:
@dalthviz, this is ready for review. Please check that:

* Shortcuts registered with `register_shortcut_for_widget` are shown in Preferences.
* Changing those shortcuts works immediately (i.e. without a restart) in their respective widgets.

dalthviz commented:
Gave this an initial check and seems like manually changing shortcuts is working and also shortcuts like `duplicate line down` are being displayed over the shortcuts preferences page :+1: However, when you modify a shortcut and then trigger a reset from the preferences page, the reset changes are not being honored inmediatly:

![widget_shortcuts](https://github.com/user-attachments/assets/af4947be-4168-44cd-a72d-7f53aa5f3365)

ccordoba12 commented:
> However, when you modify a shortcut and then trigger a reset from the preferences page, the reset changes are not being honored inmediatly

Thanks for checking that @dalthviz! It should be fixed now.

I also took the opportunity to change the context of the Switcher shortcut actions to be `_` because they are actually global shortcuts.

dalthviz commented on spyder/config/manager.py:
```diff
@@ -703,7 +723,12 @@ def set_shortcut(self, context, name, keystr, plugin_name=None):
         """
         config = self._get_shortcut_config(context, plugin_name)
-        config.set('shortcuts', context + '/' + name, keystr)
+        option = context + "/" + name
```
This could be an f-string , no?

ccordoba12 replied:
Done.

dalthviz commented on spyder/plugins/editor/widgets/codeeditor/tests/test_codeeditor.py:
```diff
@@ -693,5 +694,35 @@ def test_cell_highlight(codeeditor, qtbot):
+
+    # Check previous shortcut doesn't work
+    qtbot.keyClick(editor, Qt.Key_Down, modifier=Qt.AltModifier)
+    assert editor.toPlainText() == "bb\ncc\naa\ndd\n"
```
Could be possible to add as part of this test or in a different test a full config restart and check that the original shortcut works?

ccordoba12 replied:
Good idea. I expanded this test to include resetting shortcuts.

dalthviz commented on spyder/config/manager.py:
```diff
@@ -25,18 +25,13 @@
-    'find_replace',
-    'switcher'
-]
+EXTRA_VALID_SHORTCUT_CONTEXTS = ['_', 'find_replace']
```
I guess those other contexts were previously removed and are not needed anymore, right?

ccordoba12 replied:
Correct. `array_builder` is no longer used, `console` was replaced `internal_console` (the actual plugin name) and `switcher` is now a plugin, so we don't need it as an extra context.

dalthviz reviewed (commented):
Thanks @ccordoba12 ! Gave another local check to the reset shortcuts to default values and seems like things are working as expected now. Left a couple of questions and couple of suggestions like using an f-string at somepoint and an idea to extend a test but other than that this LGTM :+1:

pep8speaks commented:
Hello @ccordoba12! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:























* In the file [`spyder/config/tests/test_user.py`](https://github.com/spyder-ide/spyder/blob/43ef6939d48385d8cc982cc6a465433376e83e65/spyder/config/tests/test_user.py):

> [Line 379:1](https://github.com/spyder-ide/spyder/blob/43ef6939d48385d8cc982cc6a465433376e83e65/spyder/config/tests/test_user.py#L379): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

ccordoba12 commented:
@dalthviz, this is ready for another review. Besides addressing your comments, I also took the opportunity to fix other problems with shortcuts I found while doing that in my last four commits.

Specifically, I:

* Fixed getting, saving and setting shortcut options with capitalized names (third-party developers could set them like that, so we need to able to process them).
* Add a test for registering widget shortcuts for external plugins (so that that doesn't break with future changes).
* Validate that shortcut options have the right format (i.e. `context/name`) when setting defaults. This will ensure that third-party plugins use shortcuts in the way expected by Spyder.

dalthviz commented on spyder/config/manager.py:
```diff
@@ -729,12 +755,18 @@ def reset_shortcuts(self):
     from qtpy.QtWidgets import QApplication, QMessageBox
 
+    # Print traceback to show error in the terminal in case it's needed
+    print(traceback.format_exc())  # spyder: test-skip
```
Shouldn't we use the logging module for this?

ccordoba12 replied:
I don't think so because this will help third-party developers to see, even without logging enabled, that their plugin config is incorrect, like this:

![imagen](https://github.com/user-attachments/assets/a9785f34-17ab-40fa-8b67-67ab8e314d7f)

dalthviz reviewed (approved):
Thanks @ccordoba12 ! Left a question regarding the usage of `print` instead of the login module in some error handling but other than that this seems like ready to be merged :+1: 

Will leave this approved so feel free to merge this is you think is ready.

ccordoba12 commented:
@meeseeksdev please backport to 6.x
